File: x-baseproject/src/main/java/com/xia/baseproject/rxhttp/api/RestService.java
package com.xia.baseproject.rxhttp.api;

import java.util.Map;

import io.reactivex.Observable;
import okhttp3.RequestBody;
import okhttp3.ResponseBody;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.FieldMap;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.Headers;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.QueryMap;
import retrofit2.http.Streaming;
import retrofit2.http.Url;

/**
 * @author xia
 * @date 2018/7/27.
 */
public interface RestService {

    //==========================//
    //         POST请求          //
    // =========================//

    @POST
    @FormUrlEncoded
    Observable<ResponseBody> post(@Url String url, @FieldMap Map<String, String> maps);

    @POST
    Observable<ResponseBody> postBody(@Url String url, @Body Object object);

    @POST
    Observable<ResponseBody> postBody(@Url String url, @Body RequestBody body);

    @POST
    @Headers({"Content-Type: application/json", "Accept: application/json"})
    Observable<ResponseBody> postJson(@Url String url, @Body RequestBody jsonBody);


    //==========================//
    //         GET请求           //
    // =========================//

    @GET
    Observable<ResponseBody> get(@Url String url, @QueryMap Map<String, String> params);


    //==========================//
    //         DELETE请求        //
    // =========================//

    @DELETE
    Observable<ResponseBody> delete(@Url String url, @QueryMap Map<String, String> maps);

    @DELETE
    Observable<ResponseBody> deleteBody(@Url String url, @Body Object object);

    @DELETE
    Observable<ResponseBody> deleteBody(@Url String url, @Body RequestBody body);

    @DELETE
    @Headers({"Content-Type: application/json", "Accept: application/json"})
    Observable<ResponseBody> deleteJson(@Url String url, @Body RequestBody jsonBody);


    //==========================//
    //         PUT请求           //
    // =========================//

    @PUT
    Observable<ResponseBody> put(@Url String url, @QueryMap Map<String, String> maps);

    @PUT
    Observable<ResponseBody> putBody(@Url String url, @Body Object object);

    @PUT
    Observable<ResponseBody> putBody(@Url String url, @Body RequestBody body);

    @PUT
    @Headers({"Content-Type: application/json", "Accept: application/json"})
    Observable<ResponseBody> putJson(@Url String url, @Body RequestBody jsonBody);


    //==========================//
    //       文件上传下载         //
    // =========================//

    @POST
    Observable<ResponseBody> uploadFiles(@Url String url, @Body RequestBody body);

    @Streaming
    @GET
    Observable<ResponseBody> downloadFile(@Url String url);
}

File: fly-http/src/main/java/com/xia/fly/http/api/RestService.kt
package com.xia.fly.http.api

import io.reactivex.Observable
import okhttp3.RequestBody
import okhttp3.ResponseBody
import retrofit2.http.*

/**
 * @author xia
 * @date 2018/7/27.
 */
interface RestService {

    //==========================//
    //         POST请求          //
    // =========================//

    @POST
    @FormUrlEncoded
    fun post(@Url url: String, @FieldMap maps: Map<String, String>): Observable<ResponseBody>

    @POST
    fun postBody(@Url url: String, @Body `object`: Any): Observable<ResponseBody>

    @POST
    fun postBody(@Url url: String, @Body body: RequestBody): Observable<ResponseBody>

    @POST
    @Headers("Content-Type: application/json", "Accept: application/json")
    fun postJson(@Url url: String, @Body jsonBody: RequestBody): Observable<ResponseBody>


    //==========================//
    //         GET请求           //
    // =========================//

    @GET
    operator fun get(@Url url: String, @QueryMap params: Map<String, String>): Observable<ResponseBody>


    //==========================//
    //         DELETE请求        //
    // =========================//

    @DELETE
    fun delete(@Url url: String, @QueryMap maps: Map<String, String>): Observable<ResponseBody>

    @DELETE
    fun deleteBody(@Url url: String, @Body `object`: Any): Observable<ResponseBody>

    @DELETE
    fun deleteBody(@Url url: String, @Body body: RequestBody): Observable<ResponseBody>

    @DELETE
    @Headers("Content-Type: application/json", "Accept: application/json")
    fun deleteJson(@Url url: String, @Body jsonBody: RequestBody): Observable<ResponseBody>


    //==========================//
    //         PUT请求           //
    // =========================//

    @PUT
    fun put(@Url url: String, @QueryMap maps: Map<String, String>): Observable<ResponseBody>

    @PUT
    fun putBody(@Url url: String, @Body `object`: Any): Observable<ResponseBody>

    @PUT
    fun putBody(@Url url: String, @Body body: RequestBody): Observable<ResponseBody>

    @PUT
    @Headers("Content-Type: application/json", "Accept: application/json")
    fun putJson(@Url url: String, @Body jsonBody: RequestBody): Observable<ResponseBody>


    //==========================//
    //       文件上传下载         //
    // =========================//

    @POST
    fun uploadFiles(@Url url: String, @Body body: RequestBody): Observable<ResponseBody>

    @Streaming
    @GET
    fun downloadFile(@Url url: String): Observable<ResponseBody>
}
